src/main/java/org/springframework/data/relational/core/conversion/RelationalEntityDeleteWriter.java

```diff
@@ -51,22 +63,41 @@ public void write(@Nullable Object id, AggregateChange<?> aggregateChange) {
 
 	private void deleteAll(AggregateChange<?> aggregateChange) {
 
-		context.referencedEntities(aggregateChange.getEntityType(), null)
-				.forEach(p -> aggregateChange.addAction(DbAction.deleteAll(p.getLeafType(), new RelationalPropertyPath(p), null)));
+		List<DbAction> actions = new ArrayList<>();
 
-		aggregateChange.addAction(DbAction.deleteAll(aggregateChange.getEntityType(), null, null));
+		context.findPersistentPropertyPaths(aggregateChange.getEntityType(), PersistentProperty::isEntity)
+				.forEach(p -> actions.add(new DbAction.DeleteAll<>(p)));
+
+		Collections.reverse(actions);
+
+		actions.forEach(aggregateChange::addAction);
+
+		DbAction.DeleteAllRoot<?> result = new DbAction.DeleteAllRoot<>(aggregateChange.getEntityType());
+		aggregateChange.addAction(result);
 	}
 
 	private <T> void deleteById(Object id, AggregateChange<T> aggregateChange) {
 
 		deleteReferencedEntities(id, aggregateChange);
 
-		aggregateChange.addAction(DbAction.delete( //
-				id, //
-				aggregateChange.getEntityType(), //
-				aggregateChange.getEntity(), //
-				null, //
-				null //
-		));
+		aggregateChange.addAction(new DbAction.DeleteRoot<>(aggregateChange.getEntityType(), id));
```

Comment:
`AggregateChange` is mutated in deeply nested methods. We should rather return a collection of `DbAction<T>` and assemble `AggregateChange` in the top-most method.